### PR TITLE
Repository Handling Updates: `install` and `remove` commands

### PR DESCRIPTION
**Pull request by d3sch41n**
Improve handling of multiple TTP repositories - this is crucial to allow organizations
to maintain internal TTP repositories with their own secret sauce, and also facilitates working with forks. 

The new features in this PR are demonstrated as follows: 

1. Install TTPs from a new repository

Fork the ForgeArmory repository, then install your fork (substitute your username for mine)

```
./ttpforge install repo --name mine https://github.com/d3sch41n/ForgeArmory 
```

You will now see your new repository listed when you run `ttpforge list repos`

image

2. List TTPs from your new repository 

The command below will list the new TTPs

```
./ttpforge list ttps --repo mine 
```

3. Run a TTP from your new repository 

Use the `repo//my/ttp` syntax as follows:

```
./ttpforge run \
  mine//examples/args/define-args.yaml \
  --arg a_message="hello" \
  --arg a_number=1337
``` 

4. Remove the new repository 

This will clean up the new repository now that we don't need it anymore:

```
./ttpforge remove repo mine 
```
Other changes: 

* the `config.yaml` is no longer required unless you wish to use sub-TTPs or repository-scoped TTP references (repo//my/ttp)
* regular users most likely want to run `ttpforge init` and get the default config under their home - we warn about this now:

image

* However, if you are executing TTPs directly from a file path (which happens in some deployment scenarios without a config.yaml) then it will still run after the warning, as desired:

image

### CODE REVIEW

File: cmd/config.go [hunk 1]
```diff
@@ -21,44 +21,75 @@ package cmd
...
-	if err = os.MkdirAll(defaultConfigDir, 0700); err != nil {
-		return "", err
-	}
+	return repos.NewRepoCollection(fsys, cfg.RepoSpecs, cfgDir)
+}
 
-	if err := afero.WriteFile(f, defaultConfigPath, []byte(defaultConfigContents), 0600); err != nil {
-		return "", err
+// save() writes the current config back to its file - used by `install“ command
+func (cfg *Config) save() error {
```
Comment (d3sch41n) [inline cmd/config.go:83]: this new function (used to save config after repo changes) is the key change for this file - everything else is just cut-and-paste to organize a bit

File: cmd/init.go [hunk 2]
```diff
@@ -20,10 +20,24 @@ THE SOFTWARE.
...
+	"fmt"
+	"os"
+	"path/filepath"
+
 	"github.com/facebookincubator/ttpforge/pkg/logging"
+	"github.com/spf13/afero"
 	"github.com/spf13/cobra"
 )
 
+func copyEmbeddedConfigToPath(configFilePath string) error {
```
Comment (d3sch41n) [inline cmd/init.go:32]: Now. you must explicitly run `init` if you want the embedded config to get copied - this makes TTPForge compatible with deployment scenarios in which you don't want a `config.yaml`

File: cmd/list_ttps_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	}{
+		{
+			name: "no-filter",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			rc := cmd.BuildRootCommand()
+			rc.SetArgs([]string{"list", "ttps", "-c", testConfigFilePath})
```
Comment (d3sch41n) [inline cmd/list_ttps_test.go:45]: I need to do some surgery on the stdout capture handling for testing (tricky to do it in a truly thread safe manner, cobra has an idiomatic way to do it that i'll implement) - will add assertions against stdout in separate PR

File: cmd/run_test.go [hunk 4]
```diff
@@ -53,3 +53,31 @@ func TestRun(t *testing.T) {
 		})
 	}
 }
+
+// TestRunWithoutConfig verifies
+// that running TTPs still works without a top-level config file
+func TestRunWithoutConfig(t *testing.T) {
+	testCases := []struct {
```
Comment (d3sch41n) [inline cmd/run_test.go:60]: important to be able to do this for certain deployment scenarios - now has a test case

File: cmd/test-resources/test-config.yaml [hunk 5]
```diff
@@ -2,3 +2,5 @@
 repos:
   - name: test-repo
```
Comment (d3sch41n) [inline cmd/test-resources/test-config.yaml:3]: have been burned by under-testing the "this is a real Fs not afero in memory" branches a couple times, so continuing  approach in these E2E cmd-level tests of using real file system

File: pkg/repos/repo.go [hunk 6]
```diff
@@ -47,15 +47,15 @@ const (
 // in order to add a given repository folder to
 // the TTPForge search path for TTPs, templates, etc
 type Spec struct {
-	Name string     `yaml:"name"`
-	Path string     `yaml:"path"`
-	Git  *GitConfig `yaml:"git"`
+	Name string    `yaml:"name"`
+	Path string    `yaml:"path"`
+	Git  GitConfig `yaml:"git"`
```
Comment (d3sch41n) [inline pkg/repos/repo.go:52]: some subtle pointer bugs were possible with these - took pointers out

File: pkg/repos/repo_collection.go [hunk 7]
```diff
@@ -49,21 +49,19 @@ type repoCollection struct {
...
 // fsys: base file system (used for unit testing)
 // specs: a list of repo.Spec entries (usually from the config file)
-// workdirSupport: true if the provided fsys supports workdirs
 //
 // **Returns:**
 //
 // RepoCollection: assembled RepoCollection, or nil if there was an error
 // error: an error if there is a problem
-func NewRepoCollection(fsys afero.Fs, specs []Spec, workdirSupport bool) (RepoCollection, error) {
+func NewRepoCollection(fsys afero.Fs, specs []Spec, basePath string) (RepoCollection, error) {
```
Comment (d3sch41n) [inline pkg/repos/repo_collection.go:57]: `basePath` parameter added to simplify code and close several classes of bugs related to absPath handling

**Comment by d3sch41n** [comment]
Fixing last test issue once I get some sleep

**Comment by d3sch41n** [comment]
Issue is from `Conf` being a global variable - going to finally take care of that and scope it to an individual root command instance

**Review by l50 — CHANGES_REQUESTED** [review]
There are a few changes that need to be made to land - see my comments on the review I did.

File: cmd/root.go [hunk 8]
```diff
@@ -100,42 +77,52 @@ TTPForge is a Purple Team engagement tool to execute Tactics, Techniques, and Pr
...
+	if Conf.repoCollection, err = Conf.loadRepoCollection(); err != nil {
+		return err
 	}
-	rc, err := repos.NewRepoCollection(fsys, Conf.RepoSpecs, true)
-	cobra.CheckErr(err)
-	Conf.repoCollection = rc
 
-	err = logging.InitLog(logConfig)
-	cobra.CheckErr(err)
+	// setup loggign
```
[the remark this answers is not in the document]
Comment (d3sch41n) [reply cmd/root.go:126]: fixed

File: cmd/list_ttps_test.go (hunk 3 above)
Comment (d3sch41n) [reply cmd/list_ttps_test.go]: done https://github.com/facebookincubator/TTPForge/issues/315

**Comment by d3sch41n** [comment]
@l50 thank you for comments - I implemented all of them

File: pkg/repos/repo_collection.go
Comment (l50) [inline pkg/repos/repo_collection.go:1]: @d3sch41n one straggler here with an underscore. Once that's fixed, LGTM!

**Review by l50 — CHANGES_REQUESTED** [review]
@d3sch41n There was one straggler that I commented on. Once that's fixed, LGTM!

**Comment by d3sch41n** [comment]
@l50 done

**Review by l50 — APPROVED** [review]
Awesome work, approved!

**Pull request merged**